modules/gateways/Invoice/lib/Invoice.class.php

```diff
@@ -289,24 +289,22 @@ public function send_invoice($download = false)
         // Create the PDF
         if ($request->requestParamIsSet('html')) {
             echo wp_kses($content, AllowedTags::getWithFormTags());
-        } else {
-            // only load dompdf if nobody else has yet...
-            if (! class_exists('Dompdf\Dompdf')) {
-                require_once(EE_THIRD_PARTY . 'dompdf/src/Autoloader.php');
-                Dompdf\Autoloader::register();
-            }
-            $options = new Dompdf\Options();
-            $options->set('isRemoteEnabled', true);
-            $options->set('isJavascriptEnabled', false);
-            if (defined('DOMPDF_FONT_DIR')) {
-                $options->setFontDir(DOMPDF_FONT_DIR);
-                $options->setFontCache(DOMPDF_FONT_DIR);
-            }
+            exit(0);
+        }
+        $pdf_messanger = new EE_Pdf_messenger();
+        // dompdf options
+        $options = $pdf_messanger->get_dompdf_options();
+        if ($options instanceof Dompdf\Options) {
             $dompdf = new Dompdf\Dompdf($options);
             $dompdf->loadHtml($content);
             $dompdf->render();
             $dompdf->stream($invoice_name . ".pdf", ['Attachment' => $download]);
+            exit(0);
         }
```

Comment:
this logic is relatively the same as some of the lines in `EE_Pdf_messenger::do_pdf()` and could be extracted into its own method as well. Maybe with parameters for the following:

- `Dompdf\Options` 
- file name
- boolean $attachment

a potentially _better_ idea is to extract ALL of the logic that deals directly with `Dompdf` into some kind of PDF adapter class, maybe located in a new `core/services/helpers/pdf` folder (open to alternate suggestions), so that ALL of the `Dompdf` specific logic is contained in its own class and decoupled from everything else. This would make it easier in the future to update and/or replace it with another library.

This adapter could then be imported into any class that needs to generate a PDF, which i _think_ would only be:

- `core/libraries/messages/messenger/EE_Pdf_messenger.class.php`
- `modules/gateways/Invoice/lib/Invoice.class.php`

but plz search thoroughly as i may be wrong about that.

As a general rule for class creation (keep in mind there's always exceptions) i think it's best to only use class constructors for setting up dependencies and then use another method for initializing with use case and/or request specific details. Specifically, avoid side effects in class constructors and only use them for constructing the absolute minimum required for the class to function so that a class can be constructed when and where it is most useful, but external side effects like db requests don't occur until needed. Basically, classes are more flexible if logic for specific use cases is decoupled from class construction. One obvious exception to the above is DTOs (data transfer objects) where the use case / request specific details **ARE** the essential core of the class. In those cases, passing the data to the constructor makes most sense.

And my apologies for sending you off in another direction. Sometimes things don't appear to me on first review because i get distracted by some other issues. And plz keep in mind that the above is in no way a criticism of your work, as i realize you are working with code that someone else created. My interest now is to improve the existing code and fix any outstanding issues it may have.